### PR TITLE
refactor(core): Convert loadBalancers view to React

### PR DESCRIPTION
**Pull request by jrsquared**
Convert all the load balancers view and down to react, except for instances, which is interesting enough to wait until all things that depend on it are react.

One change that is out of the load balancers box is bringing it `lodash-decorators` for a much nicer way of debouncing es6 class functions. Changed all the cases we were debouncing these.

@anotherchrisberry @christopherthielen @danielpeach  PTAL (pull down and test would be good too, as a sanity check, especially wrt servergroup and instance updates in the load balancers view)

### CODE REVIEW

File: app/scripts/modules/core/cloudProvider/providerSelection/providerSelection.service.ts [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+    });
+  }
+}
+
+export let providerSelectionService: ProviderSelectionService;
+export const PROVIDER_SELECTION_SERVICE = 'spinnaker.cloudProvider.providerSelection.service';
+module (PROVIDER_SELECTION_SERVICE, [
+  ACCOUNT_SERVICE,
+  CLOUD_PROVIDER_REGISTRY,
+]).factory('providerSelectionService', ($uibModal: IModalService, $q: IQService, accountService: AccountService, cloudProviderRegistry: CloudProviderRegistry) =>
```
Comment (danielpeach) [inline app/scripts/modules/core/cloudProvider/providerSelection/providerSelection.service.ts]: Any reason to use a factory over a service here? Also, the way you've invoked the class, I don't think you need `'ngInject'` above.

File: app/scripts/modules/core/cloudProvider/providerSelection/providerSelection.service.ts [hunk 2]
```diff
@@ -0,0 +1,61 @@
+import { Application } from 'core/application/application.model';
+import { AccountService } from './../../account/account.service';
```
Comment (danielpeach) [inline app/scripts/modules/core/cloudProvider/providerSelection/providerSelection.service.ts]: Extra period.

File: app/scripts/modules/core/domain/IComponent.ts [hunk 3]
```diff
@@ -0,0 +1,5 @@
+import { IEntityTags } from './IEntityTags';
+
+export interface IComponent {
```
Comment (danielpeach) [inline app/scripts/modules/core/domain/IComponent.ts]: This sounds like a member of the `IComponentOptions` or `IComponentController` family. What about `IEntity`?

File: app/scripts/modules/core/cloudProvider/providerSelection/providerSelection.service.ts (hunk 1 above)
Comment (jrsquared) [inline app/scripts/modules/core/cloudProvider/providerSelection/providerSelection.service.ts]: Probably not? I just converted as is...

Comment (jrsquared) [inline app/scripts/modules/core/cloudProvider/providerSelection/providerSelection.service.ts]: Probably not? I just converted as is...

File: app/scripts/modules/core/domain/IComponent.ts (hunk 3 above)
Comment (jrsquared) [inline app/scripts/modules/core/domain/IComponent.ts]: Good call.

File: app/scripts/modules/core/instance/instances.component.ts [hunk 4]
```diff
@@ -106,6 +106,10 @@ export class InstancesController implements IComponentController {
     this.$scope.$watch('highlight', () => this.renderInstances());
   }
 
+  public $onChanges(): void {
```
Comment (anotherchrisberry) [inline app/scripts/modules/core/instance/instances.component.ts]: careful with this - it will actually get called twice when the component renders.

Should probably be rewritten as:
```typescript

interface IOnChanges extends IOnChangesObject {
  instances: IChangesObject<Instance[]>;
  highlight: IChangesObject<string>;
}
...
  public $onChanges(changes: IOnChanges): void {
    if (changes.instances && !changes.instances.isFirstChange()) {
      this.renderInstances();
    }
  }
```
This will avoid an initial extra string build in the render method and shave some time off the paint (try it on a very large edge service to get a feel for the performance difference).

You could make a few other changes to the `renderInstances` method that may or may not make some difference in performance:
1. instead of `sortBy(this.instances, 'launchTime')`, skip the lodash: 
  `const instances = (this.instances || []).sort((a, b) => a.launchTime - b.launchTime);`
2. instead of constantly calling `$state.includes` to see if the instance is active, grab the `provider` and `instanceId` from `$stateParams`...okay, fine, here is what I did locally:
```typescript
private renderInstances(): void {
    this.activeInstance = null;
    const instances = (this.instances || []).sort((a, b) => a.launchTime - b.launchTime);
    const showingDetails = this.$state.current.name.endsWith('.instanceDetails'),
          showingId = this.$stateParams.instanceId,
          showingProvider = this.$stateParams.provider;

    const innerHtml = '<div class="instances">' + instances.map((instance) => {
        const id = instance.id;
        let activeClass = '';

        if (showingDetails && showingId === instance.id && showingProvider === instance.provider) {
          activeClass = ' active';
          this.activeInstance = { instanceId: instance.id, provider: instance.provider };
        }
        if (this.highlight === id) {
          activeClass += ' highlighted';
        }

        return '<a title="' + id +
          '" data-provider="' + instance.provider +
          '" data-toggle="tooltip" data-instance-id="' + id +
          '" class="instance health-status-' + instance.healthState + activeClass + '"></a>';
      }).join('') + '</div>';

    if (innerHtml !== this.$element.get(0).innerHTML) {
      this.removeTooltips();
      this.$element.get(0).innerHTML = innerHtml;
    }
  }
```
That should cut the render for very large services by about half.

File: tsconfig.json [hunk 5]
```diff
@@ -26,7 +26,7 @@
     "skipLibCheck": true,
     "sourceMap": true,
     "strictNullChecks": false, // should really get to a place where we can turn this on
-    "target": "es5",
+    "target": "es6",
```
Comment (anotherchrisberry) [inline tsconfig.json]: nice

File: app/scripts/modules/core/cluster/filter/clusterFilter.service.ts [hunk 6]
```diff
@@ -469,10 +471,12 @@ export class ClusterFilterService {
...
+export let clusterFilterService: ClusterFilterService;
 export const CLUSTER_FILTER_SERVICE = 'spinnaker.core.cluster.filter.service';
 module(CLUSTER_FILTER_SERVICE, [
   require('angular-ui-router').default,
   CLUSTER_FILTER_MODEL,
   require('./multiselect.model'),
   require('../../filterModel/filter.model.service'),
-]).service('clusterFilterService', ClusterFilterService);
+]).service('clusterFilterService', ClusterFilterService)
+  .run(($injector: any) => clusterFilterService = <ClusterFilterService>$injector.get('clusterFilterService'));
```
Comment (jrsquared) [inline app/scripts/modules/core/cluster/filter/clusterFilter.service.ts]: Happy to change the way we make services importable, but if we do, let's do that wholesale in a different PR. I think the ReactInjector way makes sense when we don't already have ng.module.

File: app/scripts/modules/core/domain/IEntity.ts [hunk 7]
```diff
@@ -0,0 +1,5 @@
+import { IEntityTags } from './IEntityTags';
+
+export interface IEntity {
```
Comment (jrsquared) [inline app/scripts/modules/core/domain/IEntity.ts]: I like it. If we add more stuff, can generalize it back later.

File: app/scripts/modules/core/entityTag/entityUiTags.component.ts [hunk 8]
```diff
@@ -182,6 +182,23 @@ class EntityUiTagsComponent implements ng.IComponentOptions {
   `;
 }
 
+export class EntityUiTagsWrapperComponent implements IComponentOptions {
```
Comment (jrsquared) [inline app/scripts/modules/core/entityTag/entityUiTags.component.ts]: React only can map to `<` bindings, so the wrapper component takes `<` bindings and maps them down to `&` and `@`. It allows the callback functionality to work in both react and angular for the component and not have to change all the places we use the component in angular to not have to change from `"text"` to `"'text'"`.

@danielpeach used the pattern for `help-field` and I stole the idea from him.

Comment (anotherchrisberry) [inline app/scripts/modules/core/entityTag/entityUiTags.component.ts]: 👍

File: app/scripts/modules/core/cluster/filter/clusterFilter.service.ts (hunk 6 above)
Comment (anotherchrisberry) [inline app/scripts/modules/core/cluster/filter/clusterFilter.service.ts]: 👍

File: app/scripts/modules/core/healthCounts/healthCounts.less [hunk 9]
```diff
@@ -13,6 +13,11 @@
 }
 
 health-counts {
+  display: inherit; // Since health-counts wraps a react component with a .health-counts class that does all the layout now, make this component not have layout
```
Comment (jrsquared) [inline app/scripts/modules/core/healthCounts/healthCounts.less]: Ah, weird! Not this specific rule, but good catch.

File: tsconfig.json (hunk 5 above)
Comment (christopherthielen) [inline tsconfig.json]: What was the reason for switching, and do we know the possible ramifications?

File: app/scripts/modules/core/cloudProvider/providerSelection/providerSelection.service.ts [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+  public selectProvider(application: Application, feature: string): IPromise<string> {
+    return this.accountService.listProviders(application).then((providers) => {
+      let provider;
+      let reducedProviders: string[] = [];
+      if (feature) {
+        reducedProviders = providers.filter((p) => this.cloudProviderRegistry.hasValue(p, feature));
+      }
+
+      // reduce the providers to the smallest, unique collection taking into consideration the useProvider values
+      reducedProviders = [...new Set(reducedProviders.map((providerName) => {
```
Comment (christopherthielen) [inline app/scripts/modules/core/cloudProvider/providerSelection/providerSelection.service.ts]: interesting.  This is effectively `flatten`?

File: app/scripts/modules/core/cluster/filter/clusterFilter.service.ts [hunk 11]
```diff
@@ -46,7 +47,18 @@ export class ClusterFilterService {
...
+
+  private isFilterable: (sortFilter: any) => boolean = this.filterModelService.isFilterable;
+
+  public constructor(private ClusterFilterModel: any, private MultiselectModel: any,
+                     private $log: ILogService, private $stateParams: StateParams,
+                     private filterModelService: any) {
+    'ngInject';
+  }
+
+  @Debounce(25)
```
Comment (christopherthielen) [inline app/scripts/modules/core/cluster/filter/clusterFilter.service.ts]: Agreed, much nicer. But in this case I'm not sure a function that returns a value should be debounced.

File: app/scripts/modules/core/cluster/filter/clusterFilter.service.ts [hunk 12]
```diff
@@ -150,33 +154,33 @@ export class ClusterFilterService {
...
         const detail: any = {};
         detail[result.detail] = true;
-        this.ClusterFilterModel.sortFilter.detail = detail;
+        this.ClusterFilterModel.asFilterModel.sortFilter.detail = detail;
       }
       if (result.category) {
         const category: any = {};
         category[result.category] = true;
-        this.ClusterFilterModel.sortFilter.category = category;
+        this.ClusterFilterModel.asFilterModel.sortFilter.category = category;
```
Comment (christopherthielen) [inline app/scripts/modules/core/cluster/filter/clusterFilter.service.ts]: typical Justin

File: app/scripts/modules/core/cluster/filter/clusterFilter.service.ts (hunk 6 above)
Comment (christopherthielen) [inline app/scripts/modules/core/cluster/filter/clusterFilter.service.ts]: My rule is: if I already have an nginjector, I'll just use it

File: app/scripts/modules/core/delivery/filter/executionFilter.service.ts [hunk 13]
```diff
@@ -11,36 +12,34 @@ export class ExecutionFilterService {
...
+  }
 
-    this.updateExecutionGroups = debounce((application: Application) => {
-      if (!application) {
-        application = this.lastApplication;
-        if (!this.lastApplication) {
-          return null;
-        }
+  @Debounce(25)
+  public updateExecutionGroups(application: Application): IExecutionGroup[] {
```
Comment (christopherthielen) [inline app/scripts/modules/core/delivery/filter/executionFilter.service.ts]: again, debouncing a function that returns a value.  Is the return value ever used? If so, perhaps it should be memoized.  If not, maybe return value should be removed.

File: app/scripts/modules/core/entityTag/entityUiTags.component.ts (hunk 8 above)
Comment (christopherthielen) [inline app/scripts/modules/core/entityTag/entityUiTags.component.ts]: I really like this approach @danielpeach

File: app/scripts/modules/core/instance/instances.component.ts [hunk 14]
```diff
@@ -121,6 +136,16 @@ export class InstancesComponent implements IComponentOptions {
   public controller: any = InstancesController;
 }
 
+export class InstancesWrapperComponent implements IComponentOptions {
+  public bindings: any = {
+    instances: '<',
+    highlight: '<'
+  };
+  public controller: any = InstancesController;
+  public template = `<instances instances="$ctrl.instances" highlight="$ctrl.highlight"/>`;
```
Comment (christopherthielen) [inline app/scripts/modules/core/instance/instances.component.ts]: I don't understand this.  If Instances component has one way bindings, why do you need to wrap it?

File: app/scripts/modules/core/cluster/filter/clusterFilter.service.ts (hunk 11 above)
Comment (jrsquared) [inline app/scripts/modules/core/cluster/filter/clusterFilter.service.ts]: I agree. I'll remove the return. I should have just removed it when I first converted it, but then what would you have to comment about?

File: tsconfig.json (hunk 5 above)
Comment (jrsquared) [inline tsconfig.json]: The new version of typescript chokes on `Set` if target is es5.

File: app/scripts/modules/core/cloudProvider/providerSelection/providerSelection.service.ts (hunk 10 above)
Comment (jrsquared) [inline app/scripts/modules/core/cloudProvider/providerSelection/providerSelection.service.ts]: Possibly? This is just copypasta from the old `js` service.

File: app/scripts/modules/core/delivery/filter/executionFilter.service.ts (hunk 13 above)
Comment (anotherchrisberry) [inline app/scripts/modules/core/delivery/filter/executionFilter.service.ts]: The return value is just there to simplify testing.

This is not something we want to memoize, as its main purpose is to mutate the data inside the execution filter model.

File: app/scripts/modules/core/loadBalancer/LoadBalancerPod.tsx [hunk 15]
```diff
@@ -0,0 +1,68 @@
...
+  showServerGroups: boolean,
+  showInstances: boolean
+}
+
+interface IState {
+}
+
+@autoBindMethods
+export class LoadBalancerPod extends React.Component<IProps, IState> {
+  constructor(props: IProps) {
```
Comment (christopherthielen) [inline app/scripts/modules/core/loadBalancer/LoadBalancerPod.tsx]: This constructor seems unnecessary

**Comment by christopherthielen** [comment]
Didn't pull it down, but LGTM

File: app/scripts/modules/core/delivery/filter/executionFilter.service.ts (hunk 13 above)
Comment (jrsquared) [inline app/scripts/modules/core/delivery/filter/executionFilter.service.ts]: Ripping out the return.

File: app/scripts/modules/core/instance/instances.component.ts (hunk 14 above)
Comment (jrsquared) [inline app/scripts/modules/core/instance/instances.component.ts]: I... I have no idea. 🤔

Comment (jrsquared) [inline app/scripts/modules/core/instance/instances.component.ts]: OH, there was a weird angular bug when trying to load the instances component and I wasn't sure if there was some name collision, so that was an attempt to fix it that I never reverted.

Comment (jrsquared) [inline app/scripts/modules/core/instance/instances.component.ts]: Great catch.

File: app/scripts/modules/core/loadBalancer/LoadBalancerPod.tsx (hunk 15 above)
Comment (jrsquared) [inline app/scripts/modules/core/loadBalancer/LoadBalancerPod.tsx]: It is.

File: app/scripts/modules/core/instance/instances.component.ts (hunk 14 above)
Comment (jrsquared) [inline app/scripts/modules/core/instance/instances.component.ts]: Spoke too soon, it is needed so the angular2react has something to wrap. Since the instances component builds itself on the fly, there's nothing for React to wrap.

File: app/scripts/modules/core/cloudProvider/providerSelection/providerSelection.service.ts (hunk 10 above)
Comment (jrsquared) [inline app/scripts/modules/core/cloudProvider/providerSelection/providerSelection.service.ts]: Yeah, uniq is way cleaner too. Good call.

File: app/scripts/modules/core/instance/instances.component.ts (hunk 14 above)
Comment (christopherthielen) [inline app/scripts/modules/core/instance/instances.component.ts]: okey dokey
